sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -851,6 +852,30 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("SPARK-47972: Cast expression limitation for collations") {
+    checkError(
+      exception = intercept[ParseException]
+        (sql("SELECT cast(1 as string collate unicode)")),
+      errorClass = "UNSUPPORTED_DATATYPE",
+      parameters = Map(
+        "typeName" -> toSQLType(StringType("UNICODE"))),
+      context =
+        ExpectedContext(fragment = s"cast(1 as string collate unicode)", start = 7, stop = 39)
+    )
+
+    checkError(
+      exception = intercept[ParseException]
+        (sql("SELECT 'A' :: string collate unicode")),
+      errorClass = "UNSUPPORTED_DATATYPE",
+      parameters = Map(
+        "typeName" -> toSQLType(StringType("UNICODE"))),
+      context = ExpectedContext(fragment = s"'A' :: string collate unicode", start = 7, stop = 35)
+    )
+
+    checkAnswer(sql(s"SELECT cast(1 as string)"), Seq(Row("1")))
+    checkAnswer(sql(s"SELECT cast('A' as string)"), Seq(Row("A")))
+  }
```

Comment:
just wondering, how does this cope with something like:

```
withSQLConf(SqlApiConf.DEFAULT_COLLATION -> "UNICODE") {
  checkAnswer(sql(s"SELECT cast(1 as string)"), Seq(Row("1")))
}
```

Reply:
I can add a test for this, but it should return a string "1" collated with UNICODE collation. The aim of this PR is to block usage of `STRING COLLATE collation_name` in cast, all previous usage for default collation stay the same.